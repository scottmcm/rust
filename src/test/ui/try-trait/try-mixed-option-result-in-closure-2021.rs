// edition:2021

#![allow(dead_code, unused)]

// This pattern was found in the wild in rust-analyzer: https://github.com/rust-analyzer/rust-analyzer/pull/7735

pub fn lookup<T>(slice: &[T], indexes: impl Iterator<Item = usize>) -> Option<Vec<&T>> {
    let values = indexes //~ ERROR the `?` operator can only be used in a function that
        .map(|i| {
            let value = slice.get(i)?; //~ ERROR the `?` operator can only be used in a closure that
            Ok(value)
        })
        .collect::<Result<_, _>>()?;
    Some(values)
}

fn main() {
}
