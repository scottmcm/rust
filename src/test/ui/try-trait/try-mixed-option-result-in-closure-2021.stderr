error[E0277]: the `?` operator can only be used in a closure that returns `Result` or `Option` (or another type that implements `FromResidual`)
  --> $DIR/try-mixed-option-result-in-closure-2021.rs:10:25
   |
LL |           .map(|i| {
   |  ______________-
LL | |             let value = slice.get(i)?;
   | |                         ^^^^^^^^^^^^^ cannot use the `?` operator in a closure that returns `Result<&T, _>`
LL | |             Ok(value)
LL | |         })
   | |_________- this function should return `Result` or `Option` to accept `?`
   |
   = help: the trait `FromResidual<Option<!>>` is not implemented for `Result<&T, _>`
   = note: required by `from_residual`

error[E0277]: the `?` operator can only be used in a function that returns `Result` or `Option` (or another type that implements `FromResidual`)
  --> $DIR/try-mixed-option-result-in-closure-2021.rs:8:18
   |
LL |  / pub fn lookup<T>(slice: &[T], indexes: impl Iterator<Item = usize>) -> Option<Vec<&T>> {
LL |  |     let values = indexes
   |  |__________________^
LL | ||         .map(|i| {
LL | ||             let value = slice.get(i)?;
LL | ||             Ok(value)
LL | ||         })
LL | ||         .collect::<Result<_, _>>()?;
   | ||___________________________________^ cannot use the `?` operator in a function that returns `Option<Vec<&T>>`
LL |  |     Some(values)
LL |  | }
   |  |_- this function should return `Result` or `Option` to accept `?`
   |
   = help: the trait `FromResidual<Result<!, _>>` is not implemented for `Option<Vec<&T>>`
   = note: required by `from_residual`

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0277`.
